sui/src/rest_gateway.rs

```diff
@@ -0,0 +1,205 @@
+// Copyright (c) 2022, Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use move_core_types::identifier::Identifier;
+use move_core_types::language_storage::TypeTag;
+use serde::de::DeserializeOwned;
+use serde_json::{json, Value};
+
+use sui_core::gateway_state::gateway_responses::TransactionResponse;
+use sui_core::gateway_state::GatewayAPI;
+use sui_types::base_types::{encode_bytes_hex, ObjectID, ObjectRef, SuiAddress};
+use sui_types::messages::{Transaction, TransactionData};
+use sui_types::object::ObjectRead;
+
+use crate::rest_gateway::requests::{
+    CallRequest, MergeCoinRequest, PublishRequest, SplitCoinRequest,
+};
+use crate::rest_gateway::responses::{NamedObjectRef, ObjectResponse, TransactionBytes};
+use async_trait::async_trait;
+pub mod requests;
+pub mod responses;
+
+pub struct RestGatewayClient {
+    pub url: String,
+}
+#[async_trait]
+#[allow(unused_variables)]
+impl GatewayAPI for RestGatewayClient {
+    async fn execute_transaction(
+        &mut self,
+        tx: Transaction,
+    ) -> Result<TransactionResponse, anyhow::Error> {
+        let url = format!("{}/api/execute_transaction", self.url);
+        let data = tx.data.to_base64();
+        let sig_and_pub_key = format!("{:?}", tx.tx_signature);
+        let split = sig_and_pub_key.split('@').collect::<Vec<_>>();
+        let signature = split[0];
+        let pub_key = split[1];
+
+        let body = json!({
+            "unsigned_tx_base64" : data,
+            "signature" : signature,
+            "pub_key" : pub_key
+        });
+        Ok(Self::post(url, body).await?)
+    }
+
+    async fn transfer_coin(
```

Comment:
Maybe because this is WIP, but just wanted to confirm that we need user signature for this and many other methods right, just like the `execute_transaction` function, right? Otherwise the REST Server cannot authenticate the request.

Reply:
all there functions (transfer_coin, move_call) now become a end point to create the TransactionData object, it doesn't alter any object or writes to the network, so signature is not required. The "real" transaction only happen when execute_transaction is called with the signature.

Reply:
Thanks for the clarification! Is there any reason why `TransactionData` cannot be constructed on the client side so that we can reduce two round trips(client <> REST server) into one, which reduces latency by 50% and improve throughput by 100%?

Reply:
we can! but only for app written in rust for now, we can provide sdk to construct the data in other language in the future. But for now a rest end point will be needed to maximise compatibility

Reply:
Got it. IMO adding an extra round trip is very expensive that we probably want to get rid of before we launch Main net. In the meanwhile, we want to avoid making breaking changes to the developer facing API(we already have Geniteam and some internal applications, and imagine we onboard more partners).  cc @lxfind 


We should try to design a forward compatible API. Something like this:

```
/// Rename the existing one
async fn get_transfer_coin_transaction_data(
	&mut self,
        signer: SuiAddress,
        object_id: ObjectID,
        gas_payment: ObjectID,
        recipient: SuiAddress,
) -> Result<TransactionData, anyhow::Error>;

/// Add a forward compatible one that construct the transaction data, sign, and execute
async fn transfer_coin(
	&mut self,
        signer: impl Signer,
        object_id: ObjectRef,
        gas_payment: ObjectRef,
        recipient: SuiAddress,
) -> Result<TransactionData, anyhow::Error>;

```

where `Signer` is a trait that contains a `sign_transaction` method and a `address()` method with a default implementation. 

Regarding support for other languages, is doing [BCS](https://mysten-labs.slack.com/archives/C02UXU76482/p1648595150923249?thread_ts=1648592827.055089&cid=C02UXU76482) encoding in other languages the main challenge? cc (@bmwill let us know if you have any suggestion on this). Also I am wondering if [neon](https://neon-bindings.com/docs/introduction/) which allows you to use Rust library in NodeJS libraries works well in practice.

This is just some thoughts, and no need to address this in this PR since it's already very large. Maybe there's better API design, and I will spend some time in the next few days thinking about it and taking inspirations from the existing web3 APIs such as ethers.js and web3.js

Reply:
IIRC we did create BCS runtimes for a number of other languages you can take a look at https://crates.io/crates/serde-generate for more information on this.